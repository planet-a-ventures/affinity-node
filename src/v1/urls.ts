/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list) for more info.
 */
export const listsUrl = (list_id?: number) =>
    list_id ? `/lists/${encodeURIComponent(list_id)}` : '/lists'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list-entry) for more info.
 */
export const listEntriesUrl = (list_id: number, list_entry_id?: number) =>
    listsUrl(list_id) + (
        list_entry_id
            ? `/list-entries/${encodeURIComponent(list_entry_id)}`
            : '/list-entries'
    )

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#rate-limit) for more info.
 */
export const rateLimitUrl = () => '/rate-limit'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#whoami) for more info.
 */
export const whoAmIUrl = () => '/auth/whoami'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#fields) for more info.
 */
export const fieldsUrl = (field_id?: number) =>
    field_id ? `/fields/${encodeURIComponent(field_id)}` : '/fields'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#field-values) for more info.
 */
export const fieldValuesUrl = (field_value_id?: number) => {
    return field_value_id
        ? `/field-values/${encodeURIComponent(field_value_id)}`
        : '/field-values'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-field-values-changes) for more info.
 */
export const fieldValueChangesUrl = () => '/field-value-changes'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#organizations) for more info.
 */
export const organizationsUrl = (organization_id?: number | 'fields') => {
    return organization_id
        ? `/organizations/${encodeURIComponent(organization_id)}`
        : '/organizations'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-organizations-fields) for more info.
 */
export const organizationFieldsUrl = () => organizationsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#persons) for more info.
 */
export const personsUrl = (person_id?: number | 'fields') => {
    return person_id ? `/persons/${encodeURIComponent(person_id)}` : '/persons'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-person-fields) for more info.
 */
export const personFieldsUrl = () => personsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#entity-files) for more info.
 */
export const entityFilesUrl = (
    entity_file_id?: number,
    is_download: boolean = false,
) => {
    return entity_file_id
        ? `/entity-files` + (is_download ? '/download' : '') +
            `/${encodeURIComponent(entity_file_id)}`
        : '/entity-files'
}
