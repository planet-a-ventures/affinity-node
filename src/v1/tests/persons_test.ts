import { assertSnapshot } from '@std/testing/snapshot'
import { afterEach, beforeEach, describe, it } from '@std/testing/bdd'

import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { Affinity } from '../index.ts'
import { getRawFixture } from './get_raw_fixture.ts'
import { apiKey, isLiveRun } from './env.ts'
import { personFieldsUrl, personsUrl } from '../urls.ts'
import type { SearchPersonsRequest } from '../persons.ts'
import { mockPagingFromAllEndpoint } from './mock_paging_from_all_endpoint.ts'

describe('persons', () => {
    let mock: MockAdapter
    let affinity: Affinity

    beforeEach(() => {
        if (!isLiveRun()) {
            mock = new MockAdapter(axios, { onNoMatch: 'throwException' })
        }
        affinity = new Affinity(apiKey() || 'api_key')
    })
    afterEach(() => {
        mock?.reset()
    })

    it('can get a specific person', async (t) => {
        const person_id = 54576635
        mock?.onGet(personsUrl(person_id)).reply(
            200,
            await getRawFixture('persons/get.raw.response.json'),
        )
        const res = await affinity.persons.get({ person_id })
        await assertSnapshot(t, res)
    })

    it('can search for persons', async (t) => {
        const request = { term: 'joscha' }
        mock?.onGet(personsUrl(), { params: request }).reply(
            200,
            await getRawFixture('persons/search.raw.response.json'),
        )
        const res = await affinity.persons.search(request)
        await assertSnapshot(t, res)
    })

    it('can search for persons with the appropriate dates', async (t) => {
        const myDate = new Date(0)
        const request: SearchPersonsRequest = {
            min_first_email_date: myDate,
            term: 'Joscha',
            with_interaction_dates: true,
        }
        mock?.onGet(personsUrl(), {
            params: {
                term: request.term,
                min_first_email_date: myDate.toISOString(),
                with_interaction_dates: true,
            },
        }).reply(
            200,
            await getRawFixture('persons/search_with_dates.raw.response.json'),
        )
        const res = await affinity.persons.search(request)
        await assertSnapshot(t, res)
    })

    it('can create a new person', async (t) => {
        const data = {
            first_name: 'Alice',
            last_name: 'Doe',
            emails: ['alice@affinity.co'],
            organization_ids: [1687449],
        }
        mock?.onPost(personsUrl()).reply(
            201,
            await getRawFixture('persons/create.raw.response.json'),
        )
        const res = await affinity.persons.create(data)
        await assertSnapshot(t, res)
    })

    it('can update a person', async (t) => {
        const data = {
            person_id: 860197,
            first_name: 'Allison',
            emails: [
                'alice@affinity.co',
                'allison@example.com',
                'allison@gmail.com',
            ],
        }
        mock?.onPut(personsUrl(data.person_id)).reply(
            200,
            await getRawFixture('persons/update.raw.response.json'),
        )
        const res = await affinity.persons.update(data)
        await assertSnapshot(t, res)
    })

    it('can delete a person', async (t) => {
        const person_id = 860197
        mock?.onDelete(personsUrl(person_id)).reply(200, {
            success: true,
        })
        const res = await affinity.persons.delete({ person_id })
        await assertSnapshot(t, res)
    })

    it('can get global person fields', async (t) => {
        mock?.onGet(personFieldsUrl()).reply(
            200,
            await getRawFixture('persons/get_fields.raw.response.json'),
        )
        const res = await affinity.persons.getFields()
        await assertSnapshot(t, res)
    })

    it('iterates over all persons', async (t) => {
        const params: SearchPersonsRequest = {
            term: 'joscha',
            page_size: 1,
        }

        await mockPagingFromAllEndpoint(
            './fixtures/persons/search.raw.response.json',
            params,
            'persons',
            personsUrl,
            mock,
        )

        // {
        //     // set up pages sequentially, each referencing the one after
        //     const { default: pages } = await import(
        //         './fixtures/persons/paginated.iterator.combined.response.json',
        //         {
        //             with: {
        //                 type: 'json',
        //             },
        //         }
        //     )

        //     pages.forEach((page, i) => {
        //         const { next_page_token: previous_page_token } = pages[i - 1] ||
        //             {}
        //         const data: SearchPersonsRequest = {
        //             ...params,
        //         }
        //         if (previous_page_token) {
        //             data.page_token = previous_page_token
        //         }
        //         // console.log('Setting up page', params, page.list_entries)
        //         mock?.onGet(personsUrl(), {
        //             params: data,
        //         }).reply(
        //             200,
        //             page,
        //         )
        //     })
        // }

        let page = 0
        for await (
            const entries of affinity.persons.searchIterator(params)
        ) {
            await assertSnapshot(t, entries, {
                name: `page ${++page} of persons`,
            })
        }
    })
})
