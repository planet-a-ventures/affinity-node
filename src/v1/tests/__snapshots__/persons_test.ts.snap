export const snapshot = {};

snapshot[`persons > can get a specific person 1`] = `
{
  emails: [
    "j@p-a.com",
    "j@f.com",
    "jf@j.com",
    "j@c.com",
    "jf@a.com",
  ],
  first_name: "Joscha",
  id: 54576635,
  last_name: "Feth",
  list_entries: [
    {
      created_at: 2023-07-04T14:35:44.910Z,
      creator_id: 54530389,
      entity_id: 54576635,
      entity_type: 0,
      id: 133527223,
      list_id: 188027,
    },
    {
      created_at: 2024-05-23T12:27:09.733Z,
      creator_id: 54576635,
      entity_id: 54576635,
      entity_type: 0,
      id: 158330666,
      list_id: 247888,
    },
  ],
  organization_ids: [
    290555008,
    3584837,
    295075617,
    1517097,
    1687113,
    296715824,
  ],
  primary_email: "j@p-a.com",
  type: 1,
}
`;

snapshot[`persons > can search for persons 1`] = `
{
  next_page_token: null,
  persons: [
    {
      emails: [
        "j-m@p-a.com",
      ],
      first_name: "Joscha",
      id: 198383654,
      last_name: "Test",
      primary_email: "j-m@p-a.com",
      type: 0,
    },
    {
      emails: [
        "some@group.calendar.google.com",
      ],
      first_name: "Someone's",
      id: 191703462,
      last_name: "and Else's appointments",
      primary_email: "some@group.calendar.google.com",
      type: 0,
    },
    {
      emails: [
        "j@p-a.com",
        "j@f.com",
        "jf@j.com",
        "j@c.com",
        "jf@a.com",
      ],
      first_name: "Joscha",
      id: 54576635,
      last_name: "Feth",
      primary_email: "j@p-a.com",
      type: 1,
    },
  ],
}
`;

snapshot[`persons > can search for persons with the appropriate dates 1`] = `
{
  persons: [
    {
      emails: [
        "j@f.com",
      ],
      first_name: "Joscha",
      id: 12345678,
      interaction_dates: {
        first_email_date: 2021-05-31T16:00:37.000Z,
        first_event_date: 2021-05-20T06:30:00.000Z,
        last_chat_message_date: null,
        last_email_date: 2022-05-12T08:01:31.000Z,
        last_event_date: 2022-07-08T07:00:00.000Z,
        last_interaction_date: 2022-07-08T07:00:00.000Z,
        next_event_date: null,
      },
      interactions: {
        first_email: {
          date: 2021-05-31T16:00:37.000Z,
        },
        first_event: {
          date: 2021-05-20T06:30:00.000Z,
        },
        last_chat_message: null,
        last_email: {
          date: 2022-05-12T08:01:31.000Z,
        },
        last_event: {
          date: 2022-07-08T07:00:00.000Z,
        },
        last_interaction: {
          date: 2022-07-08T07:00:00.000Z,
        },
        next_event: null,
      },
      last_name: "Feth",
      primary_email: "j@f.com",
      type: 0,
    },
  ],
}
`;

snapshot[`persons > can create a new person 1`] = `
{
  emails: [
    "alice@affinity.co",
  ],
  first_name: "Alice",
  id: 860197,
  last_name: "Doe",
  organization_ids: [
    1687449,
  ],
  primary_email: "alice@affinity.co",
  type: 0,
}
`;

snapshot[`persons > can update a person 1`] = `
{
  emails: [
    "alice@affinity.co",
    "allison@example.com",
    "allison@gmail.com",
  ],
  first_name: "Allison",
  id: 860197,
  last_name: "Doe",
  organization_ids: [
    1687449,
  ],
  primary_email: "alice@affinity.co",
  type: 0,
}
`;

snapshot[`persons > can delete a person 1`] = `true`;

snapshot[`persons > can get global person fields 1`] = `
[
  {
    allows_multiple: true,
    dropdown_options: [],
    id: 125,
    name: "Use Case",
    value_type: 2,
  },
  {
    allows_multiple: true,
    dropdown_options: [],
    id: 198,
    name: "Referrers",
    value_type: 0,
  },
  {
    allows_multiple: false,
    dropdown_options: [],
    id: 1615,
    name: "Address",
    value_type: 5,
  },
]
`;

snapshot[`page 1 of persons 1`] = `
[
  {
    emails: [
      "j-m@p-a.com",
    ],
    first_name: "Joscha",
    id: 198383654,
    last_name: "Test",
    primary_email: "j-m@p-a.com",
    type: 0,
  },
]
`;

snapshot[`page 2 of persons 1`] = `
[
  {
    emails: [
      "some@group.calendar.google.com",
    ],
    first_name: "Someone's",
    id: 191703462,
    last_name: "and Else's appointments",
    primary_email: "some@group.calendar.google.com",
    type: 0,
  },
]
`;

snapshot[`page 3 of persons 1`] = `
[
  {
    emails: [
      "j@p-a.com",
      "j@f.com",
      "jf@j.com",
      "j@c.com",
      "jf@a.com",
    ],
    first_name: "Joscha",
    id: 54576635,
    last_name: "Feth",
    primary_email: "j@p-a.com",
    type: 1,
  },
]
`;

snapshot[`page 4 of persons 1`] = `[]`;
